src/context/values.js

```diff
@@ -0,0 +1,720 @@
+  }
+
+  /**
+   * @param {!UsedDef} used
+   * @private
+   */
+  tryUpdate_(used) {
+    // The value is not pending anymore. If any of the dependencies will remain
+    // unresolved, we will simply need to recomputed it.
+    const refreshParent = used.pending == Pending.PENDING_REFRESH_PARENT;
+
+    let newValue;
+    try {
+      newValue = this.calc_(used, refreshParent);
+    } catch (e) {
+      // This is the narrowest catch to avoid unrelated values breaking each
+      // other. The only exposure to the user-code are `needsParent` and
+      // `compute` methods in the `ContextProp`.
+      rethrowAsync(e);
+    }
+
+    // Reset pending flag. It's good to reset it after the calculation to
+    // ensure that deps are automatically covered.
+    used.pending = Pending.NOT_PENDING;
+
+    // Check if the value has been updated.
+    this.maybeUpdated_(used, newValue);
+  }
+
+  /**
+   * @param {!UsedDef} used
+   * @param {*} value
+   * @private
+   */
+  maybeUpdated_(used, value) {
+    const {prop, value: oldValue} = used;
+    const {key} = prop;
+    const usedByKey = this.usedByKey_;
+    if (
+      oldValue === value ||
+      used !== (usedByKey && usedByKey.get(key)) ||
+      !this.isConnected_()
+    ) {
+      // Either the value didn't change, or no one needs this value anymore.
+      return;
+    }
+
+    used.value = value;
+
+    // Notify subscribers.
+    const {subscribers} = used;
+    subscribers.forEach((handler) => {
+      handler(value);
+    });
+  }
+
+  /**
+   * The used value calculation algorithm.
+   *
+   * @param {!UsedDef} used
+   * @param {boolean} refreshParent
+   * @return {*|undefined} The used value.
+   * @private
+   */
+  calc_(used, refreshParent) {
+    devAssert(this.isConnected_());
+
+    const {prop, depValues} = used;
+    const {key, compute, defaultValue} = prop;
+
+    const inputsByKey = this.inputsByKey_;
+    const inputs = inputsByKey && inputsByKey.get(key);
+    const inputValues = inputs && inputs.values;
+
+    // Calculate parent value.
+    const needsParent = calcNeedsParent(prop, inputValues);
+
+    // Refresh parent if requested.
+    if (refreshParent || needsParent != Boolean(used.parentValueNode)) {
+      const newParentValueNode = needsParent
+        ? findParent(this.contextNode_, hasInput, prop, /* includeSelf */ false)
+        : null;
+      this.updateParentValueNode_(used, newParentValueNode);
+    }
+
+    // If no parent node is found, use the default value.
+    const parentValue = defined(used.parentValue)
+      ? used.parentValue
+      : needsParent && !used.parentValueNode && inputValues
+      ? defaultValue
+      : undefined;
+
+    // Calculate the "used" value.
+    let newValue = undefined;
+    const ready =
+      depValues.every(defined) && (!needsParent || defined(parentValue));
+    if (ready) {
+      const {node} = this.contextNode_;
+      if (inputValues && !compute) {
+        newValue = inputValues[0];
```

Comment:
@samouri 

> I'm not sure which parts of the old design are no longer relevant and which still are then :/. From what I can tell we are still keeping around most of the Context tree and the idea of a `ContextProp`.

We're keeping only APIs in the [latest version of the design doc](https://docs.google.com/document/d/1mEk_myi5-Q51RffLuoWU5o8UcacEBaKhiwfm32ghIdo/edit#heading=h.cj3xrdduibzd).

> My new question would be: what is the motivation for a `compute` callback? Is it mainly for reducing over multiple writers of the same property? If so, what are the pros/cons of that approach vs. more directly copying React and forcing intermediary wrappers when applicable.

There are several motivations for `compute`:
1. Just using the wrappers like in Preact can be fragile. E.g. a AND-property can be accidentally overwritten to `true`, even when the parent is `false`. Also important: unlike in React we don't have a singular component tree with a unambiguous nesting in DOM.
2. This also reduces the propagation metadata. A `CanPlay` property can depend on `CanRender` (e.g. you can't play/animate what cannot even be displayed right now). It's enough to record it once in the declaration of the `CanPlay(deps=[CanRender] compute=(input, canRender) => input && canRender)`.
3. There are some optimizations possible. The biggest cost for DOM is broadcast - i.e. trying to update subtrees in case something has changed. `compute` helps by telling us: if this node is `CanRender=false`, then it no longer depends on the parent's state and it's whole subtree will always be `CanRender=false`. It's much faster to just check the node itself to see that it's `CanRender=false`, then keep going back to the parent.

This is somewhat similar to our logic that we use today for fixed elements. To know that an element is fixed, we have to run the following code each time:

```
for (let n = element; n; n = n.parentElement) {
  if (getComputedStyle(n).position == 'fixed') {
    return true;
  }
}
return false;
```

We optimize this heavily via lots of bugs, but it's still expensive. The goal for us is to things like this a lot cheaper.

Reply:
> Unless I'm reading this wrong (highly likely), this may be more confusing than "last writer wins" in the "no compute function included" case. It looks like its "first added setter that is still around wins".

I went back and forth on this. My main motivation for using the "first added input" is that cases like this are easier to spot than the "last added". The properties that cannot be computed, many inputs is more of a bug than an intention. So the easier it is to spot it, the better. This is pretty similar when you have CSS stylesheet with multiple matching selectors with the same specificity that provide a different value of, e.g. `font-size`. In CSS the last one is used b/c there's a notion of ordering in CSS. But for ContextNode there's no real ordering - it's always async.

Reply:
I could also add a warning when two inputs are set for props that do not anticipate it.